satellite/satellitedb/overlaycache.go

```diff
@@ -33,112 +33,166 @@ type overlaycache struct {
 			criteria.ExcludedIDs = append(criteria.ExcludedIDs, n.Id)
 			criteria.ExcludedNetworks = append(criteria.ExcludedNetworks, n.LastNet)
 		}
-		if len(nodes) == count {
+		if len(nodes) == totalCount {
 			break
 		}
 	}
-
 	return nodes, nil
 }
 
-func (cache *overlaycache) SelectNewStorageNodes(ctx context.Context, count int, criteria *overlay.NodeCriteria) (nodes []*overlay.NodeDossier, err error) {
-	defer mon.Task()(&ctx)(&err)
+func buildConditions(ctx context.Context, criteria *overlay.NodeCriteria, isNewNodeQuery bool) (query string, args []interface{}, err error) {
 
 	nodeType := int(pb.NodeType_STORAGE)
 
-	safeQuery := `
-		WHERE disqualified IS NULL
-		AND suspended IS NULL
-		AND exit_initiated_at IS NULL
-		AND type = ?
-		AND free_disk >= ?
+	// initiated with reputable node conditions
+	reputationCondition := `
+	AND total_audit_count >= ?
+	AND total_uptime_count >= ?`
+
+	if isNewNodeQuery {
+		reputationCondition = `
 		AND (total_audit_count < ? OR total_uptime_count < ?)
-		AND last_contact_success > ?`
-	args := append(make([]interface{}, 0, 10),
-		nodeType, criteria.FreeDisk, criteria.AuditCount, criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
+		`
+	}
+
+	safeNodeQuery := `
+	WHERE disqualified IS NULL
+	AND suspended IS NULL
+	AND exit_initiated_at IS NULL
+	AND type = ?
+	AND free_disk >= ?
+	` + reputationCondition + `
+	AND last_contact_success > ?`
+	args = append(make([]interface{}, 0),
+		nodeType, criteria.FreeDisk, criteria.AuditCount,
+		criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
 
 	if criteria.MinimumVersion != "" {
 		v, err := version.NewSemVer(criteria.MinimumVersion)
 		if err != nil {
-			return nil, Error.New("invalid node selection criteria version: %v", err)
+			return "", nil, Error.New("invalid node selection criteria version: %v", err)
 		}
-		safeQuery += `
-			AND (major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?))))
-			AND release`
+		safeNodeQuery += `
+				AND (major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?))))
+				AND release`
 		args = append(args, v.Major, v.Major, v.Minor, v.Minor, v.Patch)
 	}
 
-	if !criteria.DistinctIP {
-		nodes, err = cache.queryNodes(ctx, criteria.ExcludedIDs, count, safeQuery, args...)
-		if err != nil {
-			return nil, err
-		}
-		return nodes, nil
-	}
-
-	for i := 0; i < 3; i++ {
-		moreNodes, err := cache.queryNodesDistinct(ctx, criteria.ExcludedIDs, criteria.ExcludedNetworks, count-len(nodes), safeQuery, criteria.DistinctIP, args...)
-		if err != nil {
-			return nil, err
-		}
-		for _, n := range moreNodes {
-			nodes = append(nodes, n)
-			criteria.ExcludedIDs = append(criteria.ExcludedIDs, n.Id)
-			criteria.ExcludedNetworks = append(criteria.ExcludedNetworks, n.LastNet)
-		}
-		if len(nodes) == count {
-			break
+	safeExcludeNodes := ""
+	if len(criteria.ExcludedIDs) > 0 {
+		safeExcludeNodes = ` AND id NOT IN (?` + strings.Repeat(", ?", len(criteria.ExcludedIDs)-1) + `)`
```

Comment:
Let's use arrays rather than repeated `?`.

Reply:
I agree, but I think that should be a separate change. This is making a pretty significant change to pretty critical code, so I'd like to minimize the size of the diff.